lib/charms/prometheus_k8s/v0/prometheus.py

```diff
@@ -375,6 +423,62 @@ def jobs(self):
 
         return scrape_jobs
 
+    def alerts(self):
+        """Fetch alerts for all relations.
+
+        A Prometheus alert rules file consists of a list of "groups". Each
+        group consists of a list of alerts (`rules`) that are sequentially
+        executed. This method returns all the alert rules provided by each
+        related consumer charm. These rules may be used to generate a
+        separate alert rules file for each relation since the returned list
+        of alert groups are indexed by relation ID. Also for each relation ID
+        associated scrape metadata such as Juju model, UUID and application
+        name are provided so the a unique name may be generated for the rules
+        file. For each relation the structure of data returned is a dictionary
+        with four keys
+
+        - groups
+        - model
+        - model_uuid
+        - application
+
+        The value of the `groups` key is such that it may be used to generate
+        a Prometheus alert rules file directly using `yaml.dump` but the
+        `groups` key itself must be included as this is required by Prometheus,
+        for example as in `yaml.dump({"groups": alert["groups"]})`.
+
+        Currently the consumer charm only accepts a list of rules and these
+        rules are all placed into a single group, even though Prometheus itself
+        allows for multiple groups within a single alert rules file.
+
+        Returns:
+            a dictionary of alert rule groups and associated scrape
+            metadata indexed by relation ID.
+        """
+        alerts = {}
+        for relation in self._charm.model.relations[self._relation_name]:
+            if len(relation.units) == 0:
```

Comment:
This might flow better as 

```python
if len(relation.units):
    # do stuff
```

Reply:
I can certainly change it to `if not len(relation.units)` . Please see my comment below on my preference for negative conditionals.

Reply:
What I meant was adjust the flow of the function to drop the `continue`, so rather than explicitly bailing when the length is zero, just do work when it isn't

Reply:
I understand. What I am saying is by bailing out when the length is zero I do not add another level of nesting. Just my preference. Perhaps taking the Zen `Flat is better than nested` a bit too far maybe, but I think not.

Reply:
switched to `relation.units` but retained negative conditional to avoid an additional level of nesting.